snapcraft/internal/build_providers/_base_provider.py

```diff
@@ -90,5 +121,78 @@ def launch_instance(self) -> None:
     def setup_snapcraft(self) -> None:
         self.echoer.info('Setting up snapcraft in {!r}'.format(
             self.instance_name))
+        if common.is_snap():
+            self._inject_snapcraft()
+        else:
+            self._install_snapcraft()
+
+    def _install_snapcraft(self) -> None:
+        """Install snapcraft from the store."""
         install_cmd = ['sudo', 'snap', 'install', 'snapcraft', '--classic']
         self._run(install_cmd)
+
+    def _inject_snapcraft(self) -> None:
+        """Install snapcraft using local assets."""
+        # Make the snaps available to the provider
+        self._mount(mountpoint=self._SNAPS_MOUNTPOINT,
+                    dev_or_path=self._snaps_path_or_dev)
+        # TODO this should be properly fixed in snapd.
+        self.echoer.info('Waiting for pending snap auto refreshes.')
+        with contextlib.suppress(errors.ProviderExecError):
+            self._run(['sudo', 'snap', 'watch', '--last=auto-refresh'])
+        # Add the store assertion, common to all snaps.
+        self._inject_assertions([
+            ['account-key', 'public-key-sha3-384={}'.format(
+                _STORE_ASSERTION_KEY)]])
+        self.echoer.info('Setting up core')
+        self._install_snap('core')
+        self.echoer.info('Setting up snapcraft')
+        self._install_snap('snapcraft')
+
+    def _inject_assertions(self, assertions: List[List[str]]):
+        with tempfile.NamedTemporaryFile() as assertion_file:
+            for assertion in assertions:
+                assertion_file.write(
+                    subprocess.check_output(['snap', 'known', *assertion]))
+                assertion_file.write(b'\n')
+            assertion_file.flush()
+
+            self._push_file(source=assertion_file.name,
+                            destination=assertion_file.name)
+            self._run(['sudo', 'snap', 'ack', assertion_file.name])
+
+    def _inject_snap(self, snap_file_path: str, *, is_dangerous: bool,
+                     is_classic: bool) -> None:
+        # Install: will do nothing if already installed
+        args = []
+        if is_dangerous:
+            args.append('--dangerous')
+        if is_classic:
+            args.append('--classic')
+
+        self._run(['sudo', 'snap', 'install', snap_file_path] + args)
+
+    def _install_snap(self, snap_name: str) -> None:
```

Comment:
It seems like this method should swap names with `_inject_snap`; they appear to be doing the opposite things.